### DISCUSSION TITLE
How to create custom ktlint rule ?

### DISCUSSION THREAD

**Discussion by dmdbilal** [oppia/oppia-android#5331] [Q&A (Other)]
I've tried with referring some articles in net about custom ktlint but most of them are outdated and it doesn't work for me. Can someone help me with this.

**Comment by yash10019coder** [comment]
Hi, @dmdbilal what problem are you facing on this can you please elaborate?

**Comment by dmdbilal** [comment]
From the online articles reference and with the help of ChatGPT, I've locally made a custom ktlint rule and followed the steps provided. But when I run the `./gradlew ktlintCheck` the rule I created doesn't work. Here for the understanding purpose, I have worked out these in my draft project. Not in Oppia-android project.

Here the rule I created.
```
class NoInternalImportRule : Rule("no-internal-import") {
    override fun beforeVisitChildNodes(
        node: ASTNode,
        autoCorrect: Boolean,
        emit: (offset: Int, errorMessage: String, canBeAutoCorrected: Boolean) -> Unit,
    ) {
        if (node.elementType == KtStubElementTypes.IMPORT_DIRECTIVE) {
            val importDirective = node.psi as KtImportDirective
            val path = importDirective.importPath?.pathStr
            if (path != null && path.contains("internal")) {
                emit(node.startOffset, "Importing from an internal package", false)
            }
        }
    }
}
```

```
class CustomRuleSetProvider : RuleSetProviderV2("no-internal-import", NO_ABOUT) {
    override fun getRuleProviders(): Set<RuleProvider> =
        setOf(
            RuleProvider { NoInternalImportRule() },
        )
}
```

I've also created the jar file for the rule
![image](https://github.com/oppia/oppia-android/assets/91405690/28d48984-8e4c-4326-8ec4-a3cef5a4f5bb)


Also i've created a separate module `custom-ktlint-rules` for this custom rule. And i've included that in the `build.gradle` of the main `app` module. by `ktlintRuleset project(':custom-ktlint-rules')`.

**Reply by dmdbilal** [reply]
In this rule, whenever there is any path consist of the directory named `internal` it should catch it and throw an error.

eg. `import com.example.myplayground.internal.Hi`
